step-plans/step-plans-core/src/main/java/step/core/artefacts/reports/resolvedplan/ResolvedPlanBuilder.java

```diff
@@ -74,4 +82,36 @@ private ResolvedPlanNode buildTreeRecursively(String parentId, AbstractArtefact
 
         return resolvedPlanNode;
     }
+
+    /**
+     * Check first if it is a plan artefact, or a work artefact which are implicitly created at runtime (iterations, implicit session...)
+     * For work artifact, verify whether the artefact hash is known, otherwise create a new resolved plan node and attach it to the closest known parent
+     * @param artefactHash the hash of the current artefact to be checked and added if missing
+     * @param artefact the artefact referenced by the artefact hash
+     * @param parentReportNode  the parent report node (might not correspond to an actual plan node
+     * @param reportNodeCache the report node cache of the execution used to retrieve the closest parent plan's node
+     * @param <ARTEFACT> any class extending AbstractArtefact
+     */
+    public <ARTEFACT extends AbstractArtefact> void checkAndAddMissingResolvedPlanNode(String artefactHash, ARTEFACT artefact, ReportNode parentReportNode, ReportNodeCache reportNodeCache) {
+        if (!artefact.isWorkArtefact()) {
+            resolvedPlanCache.computeIfAbsent(artefactHash, n -> {
+                AbstractArtefact artefactClone = dynamicBeanResolver.cloneDynamicValues(artefact);
+                String parentPlanNodeId = findClosestParentNodeId(parentReportNode, reportNodeCache);
+                ResolvedPlanNode resolvedPlanNode = new ResolvedPlanNode(artefactClone, artefactHash, parentPlanNodeId);
+                resolvedPlanNodeAccessor.save(resolvedPlanNode);
+                return resolvedPlanNode.getId().toHexString();
+            });
+        }
+    }
+
+    private String findClosestParentNodeId(ReportNode parentReportNode, ReportNodeCache reportNodeCache) {
+        String planNodeId = resolvedPlanCache.get(parentReportNode.getArtefactHash());
+        if (planNodeId == null) {
+            ObjectId parentID = parentReportNode.getParentID();
+            if (parentID != null) {
+                planNodeId = findClosestParentNodeId(reportNodeCache.get(parentID), reportNodeCache);
```

Comment:
It shouldn't happen but we might throw a RuntimeException if we have to return null in this method i.e. if a node has no parent (root) and no corresponding planNodeId